pkg/controllermanager/controller/shoot/shoot_hibernation_control.go

```diff
@@ -56,85 +60,144 @@ func (c *Controller) shootHibernationUpdate(oldObj, newObj interface{}) {
-			return
+	for {
+		t2 := s.Schedule.Next(t1)
+		if t2.After(to) {
+			break
 		}
-		c.shootHibernationQueue.Add(key)
+		t1 = t2
 	}
+	inUTC := t1.UTC()
+	return &inUTC
 }
 
-var (
-	// TimeNow returns the current time. Exposed for testing.
-	TimeNow = time.Now
-)
-
 // NewShootHibernationReconciler creates a new instance of a reconciler which hibernates shoots or wakes them up.
 func NewShootHibernationReconciler(
-	l logrus.FieldLogger,
 	gardenClient client.Client,
 	recorder record.EventRecorder,
 ) reconcile.Reconciler {
 	return &shootHibernationReconciler{
-		logger:       l,
 		gardenClient: gardenClient,
 		recorder:     recorder,
 	}
 }
 
 type shootHibernationReconciler struct {
-	logger       logrus.FieldLogger
 	gardenClient client.Client
 	recorder     record.EventRecorder
 }
 
 func (r *shootHibernationReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
-	key := fmt.Sprintf("%s/%s", request.Namespace, request.Name)
+	log := logf.FromContext(ctx).WithName(ControllerName)
 
 	shoot := &gardencorev1beta1.Shoot{}
 	if err := r.gardenClient.Get(ctx, request.NamespacedName, shoot); err != nil {
 		if apierrors.IsNotFound(err) {
-			r.logger.Infof("Object %q is gone, stop reconciling: %v", request.Name, err)
+			log.Info("Shoot is gone, stopping reconciliation", "shoot", client.ObjectKeyFromObject(shoot))
 			return reconcile.Result{}, nil
 		}
-		r.logger.Infof("Unable to retrieve object %q from store: %v", request.Name, err)
+		log.Error(err, "Unable to retrieve shoot from store", "shoot", client.ObjectKeyFromObject(shoot))
 		return reconcile.Result{}, err
 	}
 
-	logger := r.logger.WithField("shoot-hibernation", key)
-	logger.Info("[SHOOT HIBERNATION]")
+	log = log.WithValues("shoot", client.ObjectKeyFromObject(shoot))
 
 	if shoot.DeletionTimestamp != nil {
+		log.Info("Shoot is currently being deleted, stopping reconciliation")
+		return reconcile.Result{}, nil
+	}
+	return r.reconcile(ctx, shoot, log)
+}
+
+func (r *shootHibernationReconciler) reconcile(ctx context.Context, shoot *gardencorev1beta1.Shoot, log logr.Logger) (reconcile.Result, error) {
+	schedules := getShootHibernationSchedules(shoot)
+	if schedules == nil {
+		log.Info("Hibernation schedules have been removed from shoot, stopping reconciliation")
+		return reconcile.Result{}, nil
+	}
+
+	parsedSchedules, err := parseHibernationSchedules(schedules)
+	if err != nil {
+		log.Info("Invalid hibernation schedules, stopping reconciliation")
 		return reconcile.Result{}, nil
 	}
 
-	return reconcile.Result{}, nil
+	// get the schedule which caused the current reconciliation, to check whether the shoot should be hibernated or woken up
+	mostRecentSchedule := getScheduleWithMostRecentTime(parsedSchedules, TimeNow(), shoot)
+	if mostRecentSchedule != nil {
+		patch := client.MergeFrom(shoot.DeepCopy())
+		shoot.Spec.Hibernation.Enabled = &mostRecentSchedule.Enabled
```

Comment:
Earlier, before patching the `Shoot`, we checked whether it's in `Failed` state: https://github.com/gardener/gardener/blob/471f9184f35fda756b267badb039659c6a2314e5/pkg/controllermanager/controller/shoot/shoot_hibernation_types.go#L114-L116

I think we should still do the same.